### ISSUE TITLE
Validation before "validate" button pressed

### ISSUE THREAD

**Issue by esandier** [PremierLangage/premierlangage#231]
It would be nice to have validation of user input in a different tag than `evaluator`,  and which would take place before the `validate`button is pressed in the form, giving the user instant feedback. This could be achieved by providing a `validate` tag in the `pl`, which would specify, for each input `form_something`, a name of function `validate_something` which would return a `bool`and maybe a message to be displayed next to the input field in the form.

A general principle could be that when treating the user input, the limit between `validate` and `evaluator` is that the former says wether sense can be made of the answer, and the latter says wether the answer is right or wrong. If `evaluator` cannot determine what is wrong with an answer, it means it shouldn't have been validated in the first place. In particular every `try` clause should belong to the `validate` part.

**Comment by nimdanor** [comment]
You propose that a form input should have a validate javascript fonction that checks the input before sending it the sandbox ?
Or do you propose that we write a new grader that try's the validate then if it passes the evaluator is run ?

**Comment by esandier** [comment]
I propose a new grader, because javascript cannot do symbolic math, and a typical validation would be that sympy can parse the input. 

So the javascript in playexo.js would listen to input fields, and when they are modified would send an ajax request for validation, and according to the result would paint the input field a certain color to indicate that input is not yet understood or that it is understood. 

The evaluator would only be called once input can be understood, otherwise the validate button would do nothing.

**Comment by mciissee** [comment]
It’s not easy to handle the changing state of the form elements because sometimes the teachers use external APIs that does not provide callback functions for change event and something form elements are added dynamically...

I think that the best way to validate the form is to use onBeforeSubmitPL callback.

You can do all the validation (show a message, paint html elements...) inside this function and return true or false to continue or not the evaluation.

This feature is not currently implemented in the 0.5.3 version, but it should be added to the next version with a confirmation from @nimdanor

You can test the callback functions on pl-test4.u-pem.fr @esandier by opening the file events.pl (ctrl+o to search the file)

**Comment by mciissee** [comment]
Unfortunately, you cannot use sympy functions with this proposition

**Comment by esandier** [comment]
yes, that is the main issue which prevents validation from being done in javascript. 

The use of external API's for user input is not really a problem, because in practice the end-user will use predefined elements to define his `form`.

It  is for the designer of these elements to define the code which captures the student's modifications on the element, and then calls a function which could be called `onModifyPLInput` or something like that, which would handle communication with the server.

**Comment by nimdanor** [comment]
We have an other appoach with Nicolas is adding a negative grade foir this kind of things.
If score==-2 then it's a validation problème a not a real grade.
But before we commit to a solution i would like to try my tag syntax, where the validation is one of the elments.

**Comment by esandier** [comment]
OK. But the score == -2 solution does not allow to distinguish different input fields, and still requires to submit the form. I would like a more interactive process where validation code for an input field  is invoked each time the field is updated, like in many forms found on the web.
